openmdao/components/tests/test_external_code_comp.py

```diff
@@ -610,13 +613,96 @@ def setup(self):
                     sys.executable, 'extcode_mach.py', self.input_file, self.output_file,
                 ]
 
+                # If you want to write your own string command, the code below will also work.
+                # self.options['command_apply'] = 'python extcode_mach.py mach_input.dat mach_output.dat'
+
+            def apply_nonlinear(self, inputs, outputs, residuals):
+                with open(self.input_file, 'w') as input_file:
+                    input_file.write('residuals\n')
+                    input_file.write('{}\n'.format(inputs['area_ratio'][0]))
+                    input_file.write('{}\n'.format(outputs['mach'][0]))
+
+                # the parent apply_nonlinear function actually runs the external code
+                super(MachExternalCodeComp, self).apply_nonlinear(inputs, outputs, residuals)
+
+                # parse the output file from the external code and set the value of mach
+                with open(self.output_file, 'r') as output_file:
+                    mach = float(output_file.read())
+                residuals['mach'] = mach
+
+            def solve_nonlinear(self, inputs, outputs):
+                with open(self.input_file, 'w') as input_file:
+                    input_file.write('outputs\n')
+                    input_file.write('{}\n'.format(inputs['area_ratio'][0]))
+                    input_file.write('{}\n'.format(self.options['super_sonic']))
+                # the parent apply_nonlinear function actually runs the external code
+                super(MachExternalCodeComp, self).solve_nonlinear(inputs, outputs)
+
+                # parse the output file from the external code and set the value of mach
+                with open(self.output_file, 'r') as output_file:
+                    mach = float(output_file.read())
+                outputs['mach'] = mach
+
+        group = om.Group()
+        group.add_subsystem('ar', om.IndepVarComp('area_ratio', 0.5))
+        mach_comp = group.add_subsystem('comp', MachExternalCodeComp(), promotes=['*'])
+        prob = om.Problem(model=group)
+        group.nonlinear_solver = om.NewtonSolver()
+        group.nonlinear_solver.options['solve_subsystems'] = True
+        group.nonlinear_solver.options['iprint'] = 0
+        group.nonlinear_solver.options['maxiter'] = 20
+        group.linear_solver = om.DirectSolver()
+
+        prob.setup()
+
+        area_ratio = 1.3
+        super_sonic = False
+        prob['area_ratio'] = area_ratio
+        mach_comp.options['super_sonic'] = super_sonic
+        prob.run_model()
+        assert_rel_error(self, prob['mach'], mach_solve(area_ratio, super_sonic=super_sonic), 1e-8)
+
+        area_ratio = 1.3
+        super_sonic = True
+        prob['area_ratio'] = area_ratio
+        mach_comp.options['super_sonic'] = super_sonic
+        prob.run_model()
+        assert_rel_error(self, prob['mach'], mach_solve(area_ratio, super_sonic=super_sonic), 1e-8)
+
+    def test_string_command(self):
+        import sys
```

Comment:
This new test isn't a feature test, so it doesn't need these imports. However, maybe it should be in the docs. Maybe it makes sense to use the string version as the preferred way of demonstrating how to set the commands.

Reply:
I agree that passing a string seems like a simpler option for users and should probably be shown in the docs.